src/Components/ModuleForm/ModuleForm.js

```diff
@@ -0,0 +1,310 @@
+        console.log("Remove duplicated module");
+        this.setState({
+          error: "Duplicate Module Name",
+          moduleList: [...moduleList],
+        });
+        event.preventDefault();
+        return;
+      } else {
+        this.setState({ error: null });
+      }
+
+      console.log("Add Module");
+      this.setState(
+        {
+          moduleList: [...moduleList, newModule],
+          moduleName: "",
+          grade: this.state.grade,
+          userAction: "Add",
+        },
+        () => {
+          localStorage.setItem(
+            "moduleList",
+            JSON.stringify(this.state.moduleList)
+          );
+        }
+      );
+
+      // Add logic for updating modules
+    } else if (userAction === "Update") {
+      const newModule = { moduleName, grade };
+      const updatedModuleList = moduleList.map((module) => {
+        if (module.moduleName === newModule.moduleName) {
+          return newModule;
+        } else {
+          return module;
+        }
+      });
+
+      if (updatedModuleList.length === moduleList.length) {
+        console.log("No module found");
+        this.setState({
+          error: "No module found",
+          moduleList: [...moduleList],
+        });
+        event.preventDefault();
+        return;
+      }
+
+      this.setState(
+        {
+          moduleList: updatedModuleList,
+          moduleName: "",
+          grade: "",
+          userAction: "Update",
+        },
+        () => {
+          localStorage.setItem(
+            "moduleList",
+            JSON.stringify(this.state.moduleList)
+          );
+        }
+      );
+
+      // Add logic for deleting modules
+    } else if (userAction === "Delete") {
+      const newModule = { moduleName, grade };
+      const updatedModuleList = moduleList.filter(
+        (module) => module.moduleName !== newModule.moduleName
+      );
+
+      if (updatedModuleList.length === moduleList.length) {
+        console.log("No module found");
+        this.setState({
+          error: "No module found",
+          moduleList: [...moduleList],
+        });
+        event.preventDefault();
+        return;
+      }
+
+      this.setState(
+        {
+          moduleList: [...updatedModuleList],
+          moduleName: "",
+          grade: this.state.grade,
+          userAction: "Delete",
+        },
+        () => {
+          localStorage.setItem(
+            "moduleList",
+            JSON.stringify(this.state.moduleList)
+          );
+        }
+      );
+      // Add logic for deleting modules
+      console.log("Delete Module");
+    }
+
+    event.preventDefault();
+  };
+
+  handleActionChange = (event) => {
+    this.setState({
+      userAction: event.target.value,
+    });
+    console.log("Action Change");
+    event.preventDefault();
+  };
+
+  calculateGPA = () => {
+    const { moduleList } = this.state;
+    const totalModules = moduleList.length;
+    if (totalModules === 0) {
+      return 0;
+    }
```

Comment:
```suggestion
    if (!totalModules) return 0;
```

Reply:
Is this necessary though? I guess it saves you the reduce operation below, but if totalModules is 0, then below calculation will always be 0 anyways. I guess a matter of personal choice here and slightly better runtime. Though the `reduce` loop for a few modules wouldn't be so bad in performance anyways.